### PR TITLE
[FLINK-6337][network] Remove the buffer provider from PartitionRequestServerHandler

### PR DESCRIPTION
**Pull request by zhijiangW**
Currently, `PartitionRequestServerHandler` will create a `LocalBufferPool` when the channel is registered. The `LocalBufferPool` is only used to get segment size for creating read view in `SpillableSubpartition`, and the buffers in the pool will not be used all the time, so it will waste the buffer resource of global pool.

We would like to remove the `LocalBufferPool` from the `PartitionRequestServerHandler`, and the `LocalBufferPool` in `ResultPartition` can also provide the segment size for creating sub partition view.

This modification will not effect the current behavior and will get benefits of saving buffer resources.

### CODE REVIEW

**Comment by zhijiangW** [comment]
Hey @uce , I have submitted the pull request for the issue we confirmed before.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java [hunk 1]
```diff
@@ -351,15 +351,15 @@ public void destroyBufferPool() {
...
+	public ResultSubpartitionView createSubpartitionView(int index, BufferAvailabilityListener availabilityListener) throws IOException {
 		int refCnt = pendingReferences.get();
 
 		checkState(refCnt != -1, "Partition released.");
 		checkState(refCnt > 0, "Partition not pinned.");
 
 		checkElementIndex(index, subpartitions.length, "Subpartition not found.");
 
-		ResultSubpartitionView readView = subpartitions[index].createReadView(bufferProvider, availabilityListener);
+		ResultSubpartitionView readView = subpartitions[index].createReadView(bufferPool, availabilityListener);
```
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java]: I think we can completely remove the buffer provider from the `createReadView` method:

- In `SpillableSubpartition#createReadView` we can use the segment size of the buffer pool of the spillable subpartition itself (`parent.getBufferProvider().getMemorySegmentSize()`).
- In `PipelinedSubpartition#createReadView` we don't use the argument anyways.

File: flink-tests/src/test/java/org/apache/flink/test/misc/SuccessAfterNetworkBuffersFailureITCase.java [hunk 2]
```diff
@@ -68,7 +68,6 @@ public void testSuccessfulProgramAfterFailure() {
 	
 			try {
 				runKMeans(cluster.getLeaderRPCPort());
-				fail("This program execution should have failed.");
```
Comment (uce) [inline flink-tests/src/test/java/org/apache/flink/test/misc/SuccessAfterNetworkBuffersFailureITCase.java]: Why did you remove this line? I think it's important to keep this line here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java (hunk 1 above)
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java]: Thank you for so efficient reviews! 
I agree with your suggestion. It will be more complete and pretty good. 
I will modify and submit another commit for it, and verify the tests again.

File: flink-tests/src/test/java/org/apache/flink/test/misc/SuccessAfterNetworkBuffersFailureITCase.java (hunk 2 above)
Comment (zhijiangW) [inline flink-tests/src/test/java/org/apache/flink/test/misc/SuccessAfterNetworkBuffersFailureITCase.java]: This case will fail if not remove it. 

I guess the buffer amount is not enough for this case before, because some are shared by buffer provider in handler. After remove the buffer provider from handler, the buffer is enough now. So the `runKmeans` can execute correctly, not enter catch part or fail process.

I am not familiar with this case and not sure whether my understanding is right.

**Comment by zhijiangW** [comment]
Hi @uce , I have submitted the modifications of `createReadView`, and the tests and IT have passed in my private travis.  I checked the failed test `HistoryServerTest.testFullArchiveLifecycle` on this travis and it seems no related to my pull request. I run it several times separately again and all are passed.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionTest.java [hunk 3]
```diff
@@ -68,7 +68,11 @@ public static void shutdown() {
 
 	@Override
 	ResultSubpartition createSubpartition() {
-		return new SpillableSubpartition(0, mock(ResultPartition.class), ioManager);
+		ResultPartition parent = mock(ResultPartition.class);
```
Comment (uce) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionTest.java]: Could you make this a static helper method?
```java
private static ResultPartition createMockPartition() {
   ...
}
```

File: flink-tests/src/test/java/org/apache/flink/test/misc/SuccessAfterNetworkBuffersFailureITCase.java (hunk 2 above)
Comment (uce) [inline flink-tests/src/test/java/org/apache/flink/test/misc/SuccessAfterNetworkBuffersFailureITCase.java]: I think that the idea of the test is to check that another program can be executed after one failed. The commit message introducing the test says "This test validates that task slots in co-location constraints are properly freed in the presence of failures." By removing this line, we are not testing what we actually wanted to test anymore. We should keep the line but instead decrease the number of configured buffers, for example 640 works instead of 840. That way we keep the behaviour: successful job, failed job, successful job.

Comment (zhijiangW) [inline flink-tests/src/test/java/org/apache/flink/test/misc/SuccessAfterNetworkBuffersFailureITCase.java]: Yes, I misunderstand this test and you get the key point.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionTest.java (hunk 3 above)
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/SpillableSubpartitionTest.java]: I guess you want to reuse this mocked `ResultPartition` in all the internal tests, and I notice that there already exists `createSubpartition` override method in this test, so I can modify the return type of `SpillableSubpartition` instead. Do you think so?

**Comment by zhijiangW** [comment]
Hi @uce, I have submitted the modifications for the above two issues. Welcome any comments! :-)

**Comment by uce** [comment]
Thanks! Looks good to merge now 👍 I've rebased it and pushed it to Travis here: https://travis-ci.org/uce/flink/builds/227891373

As soon as it succeeds, I'm going to merge this PR.
